tests/unit/test_credentials.py

```diff
@@ -972,5 +973,137 @@ def test_source_profile_does_not_exist(self):
+
+    def create_client_creator(self, with_response=None):
+        # Create a mock sts client that returns a specific response
+        # for assume_role_with_saml.
+        client = mock.Mock()
+        client.assume_role_with_saml.return_value = with_response or {
+            'Credentials': {
+                'AccessKeyId': 'foo',
+                'SecretAccessKey': 'bar',
+                'SessionToken': 'baz',
+                'Expiration': datetime.now(tzlocal()).isoformat(),
+            },
+        }
+        return mock.Mock(return_value=client)
+
+    def load_cred(self, profile):
+        provider = credentials.AssumeRoleWithSamlProvider(
+            lambda: {'profiles': {'saml': profile}},
+            self.create_client_creator(), cache={}, profile_name='saml',
+            username_prompter=lambda prompt: 'joe',
+            password_prompter=lambda prompt: 'secret')
+        creds = provider.load()
+        with self.assertRaises(botocore.exceptions.RefreshUnsupportedError):
+            # access_key is a property that will refresh credentials.
+            # Forms-based SAML authentication will currently raise an exception
+            creds.access_key
+
+    @mock.patch(GET, return_value=mock.Mock(
+        text='<form><input name="username"/><input name="password"/></form>'))
+    def test_assume_role_with_okta(self, _get):
+        profile = {
+            'saml_endpoint': 'http://example.com/login.asp',
+            'saml_authentication_type': 'form',
+            'saml_provider': 'okta',
+            'saml_username': 'joe',
+            'role_arn': 'arn:aws:iam::123456789012:role/joe',
+        }
+        self.load_cred(profile)
+        self._post.assert_called_with(
+            'http://example.com/login.asp',
+            data={'username': 'joe', 'password': 'secret'}, verify=True)
+
+    @mock.patch(GET, return_value=mock.Mock(text=adfs_login_form))
+    def test_assume_role_with_adfs(self, _get):
+        profile = {
+            'saml_endpoint': 'http://example.com/login.asp',
+            'saml_authentication_type': 'form',
+            'saml_provider': 'adfs',
+            'saml_username': 'joe',
+            'role_arn': 'arn:aws:iam::123456789012:role/joe',
+        }
+        self.load_cred(profile)
+        self._post.assert_called_with(
+            'http://example.com/login.asp', verify=True,
+            data={'ctl00$ContentPlaceHolder1$UsernameTextBox': 'joe',
+                  'ctl00$ContentPlaceHolder1$PasswordTextBox': 'secret'})
+
+    @mock.patch(GET, return_value=mock.Mock(text=adfs_login_form))
+    def test_unmatch_role(self, _get):
+        profile = {
+            'saml_endpoint': 'http://example.com/login.asp',
+            'saml_authentication_type': 'form',
+            'saml_provider': 'adfs',
+            'saml_username': 'joe',
+            'role_arn': 'arn:aws:iam::123456789012:role/does_not_match',
+        }
+        with self.assertRaises(botocore.exceptions.SamlError):
```

Comment:
Maybe use ``assertRaisesRegexp`` to make sure it was the iam role that caused the failure as ``SamlError`` can be thrown for a few reasons?